dpnp/dpnp_utils/dpnp_utils_linearalgebra.py

```diff
@@ -0,0 +1,358 @@
+            if x1_shape[i] != x2_shape[i]:
+                if x1_shape[i] == 1:
+                    tmp_shape[i] = x2_shape[i]
+                    # If the `x1` array is inherently 2D, there's no need to
+                    # duplicate the data for the 1-D dimension;
+                    # GEMM handles it automatically.
+                    if not x1_is_2D:
+                        x1 = dpnp.repeat(x1, x2_shape[i], axis=i)
+                elif x2_shape[i] == 1:
+                    tmp_shape[i] = x1_shape[i]
+                    if not x2_is_2D:
+                        x2 = dpnp.repeat(x2, x1_shape[i], axis=i)
+                else:
+                    raise ValueError(
+                        "arrays could not be broadcast together with remapped shapes."
+                    )
+        x1_shape = x1.shape
+        x2_shape = x2.shape
+        res_shape = tuple(tmp_shape) + (x1_shape[-2], x2_shape[-1])
+
+    # input arrays should have the proper data type
+    # and be C_CONTIGUOUS or F_CONTIGUOUS
+    dep_events_list = []
+    host_tasks_list = []
+    x1 = _get_gemm_contig_array(
+        x1, dep_events_list, host_tasks_list, gemm_dtype
+    )
+    x2 = _get_gemm_contig_array(
+        x2, dep_events_list, host_tasks_list, gemm_dtype
+    )
+
+    # calculate results
+    result = dpnp.empty(
+        res_shape,
+        dtype=gemm_dtype,
+        usm_type=res_usm_type,
+        sycl_queue=exec_q,
+    )
+    if result.size == 0:
+        pass
+    else:
+        if x1.size == 0 or x2.size == 0:
+            result.fill(0)
+        else:
+            if x1_is_2D and x2_is_2D:
+                ht_blas_ev, _ = bi._gemm(
+                    exec_q,
+                    dpnp.get_usm_ndarray(x1),
+                    dpnp.get_usm_ndarray(x2),
+                    dpnp.get_usm_ndarray(result),
+                    dep_events_list,
+                )
+            else:
+                (
+                    ht_blas_ev,
+                    ht_copy_ev,
+                    result,
+                ) = _gemm_batch_matmul(
+                    exec_q,
+                    x1,
+                    x2,
+                    result,
+                    x1_is_2D,
+                    x2_is_2D,
+                    dep_events_list,
+                )
+                host_tasks_list += ht_copy_ev
+
+            host_tasks_list.append(ht_blas_ev)
+
+    dpctl.SyclEvent.wait_for(host_tasks_list)
+
+    if squeeze_flag:
+        result = dpnp.squeeze(result)
+
+    if x1_is_2D and x2_is_2D:
+        # add new axes only if one of the input arrays
+        # was inehrently 2D
+        new_size = max(x1_ndim, x2_ndim)
+        for _ in range(new_size - 2):
+            result = result[dpnp.newaxis, :]
+
+    if gemm_dtype != res_dtype:
+        result = dpnp.astype(result, res_dtype, copy=False)
+    if out is None:
+        # If `order` was not passed as default
+        # we must copy `result` to match the passed `order`.
```

Comment:
I'm a bit confused about the comment. We are passing `copy=False` to `dpnp.array(...)`, doesn't it mean that no copy will be done?

Reply:
The comment is out of date. I updated it accordingly.